modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/SidebarPanel.js

```diff
@@ -34,38 +35,49 @@ const SidebarPanel = React.forwardRef(
 		const [isOpen, setIsOpen] = useState(true);
 		const [resourceData, setResourceData] = useState();
 
-		const getData = useCallback((fetchURL) => {
-			setIsLoading(true);
-			setResourceData(null);
+		const delay = useTimeout();
 
-			fetch(fetchURL, {
-				method: 'GET',
-			})
-				.then((response) => response.json())
-				.then((data) => {
-					setData(data, data?.error);
+		const getData = useCallback(
+			(fetchURL) => {
+				setIsLoading(true);
+				setResourceData(null);
+
+				fetch(fetchURL, {
+					method: 'GET',
 				})
-				.catch(() => {
-					setData(
-						null,
-						Liferay.Language.get('an-unexpected-error-occurred')
-					);
-				});
-		}, []);
+					.then((response) =>
+						response.headers.get('content-type').includes('json')
+							? response
+									.json()
+									.then((data) => setData(data, data?.error))
+							: response.text().then((html) => setData({html}))
+					)
+					.catch(() => {
+						setData(
+							null,
+							Liferay.Language.get('an-unexpected-error-occurred')
+						);
+					});
```

Comment:
It's not safe to call `setState()` functions from asynchronous callbacks because the component may get unmounted before the call. You have two options to fix that:

- One is to add [an `isMounted()` check](https://github.com/liferay/liferay-portal/blob/9a9b3d1bc8199ef517d1ea94ee4158ba3ae4aeda/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/hooks/useIsMounted.es.js) before these `setData()` calls (or put it inside `setData()` itself).
- Another is to use a `useStateSafe()` hook that has the `isMounted()` check built into it. I [made one of these](https://github.com/liferay/liferay-portal/blob/9a9b3d1bc8199ef517d1ea94ee4158ba3ae4aeda/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/core/hooks/useStateSafe.js) for layout-content-page-editor-web.

Given that the hook is pretty simple _and_ has already been copied into at least [one other place](https://github.com/liferay/liferay-portal/blob/9a9b3d1bc8199ef517d1ea94ee4158ba3ae4aeda/modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/hooks/useStateSafe.es.js), I think it's time to move it into frontend-js-web.

My suggestion for this one is to leave it as-is for now and move forward, and I'll create a ticket to move the hook (in fact, if I search, I might find that I already did 😂), at which point we can follow up and update these usages.

Reply:
Ticket for moving `useStateSafe()`: https://issues.liferay.com/browse/LPS-117821

Reply:
Nice, thx @wincent ! I'm watching that issue so when it's done we can use it here :)